File: clients/ts/tests/common/harness/raydiumCpmm.ts
import { venues, type VenueWindow } from "../../../src/index.js";
import { slotAt, type Fixture } from "../fixture.js";
import { key } from "../keys.js";
import { assertProgram } from "./index.js";

const { PROGRAM_ADDRESS, resolve: build } = venues.raydiumCpmm;

export function resolve(fixture: Fixture): VenueWindow {
  assertProgram(fixture, PROGRAM_ADDRESS);
  return build({
    user: slotAt(fixture, 0).pubkey,
    ammConfig: slotAt(fixture, 2).pubkey,
    pool: slotAt(fixture, 3).pubkey,
    inputTokenAccount: slotAt(fixture, 4).pubkey,
    outputTokenAccount: slotAt(fixture, 5).pubkey,
    inputVault: slotAt(fixture, 6).pubkey,
    outputVault: slotAt(fixture, 7).pubkey,
    inputTokenProgram: slotAt(fixture, 8).pubkey,
    outputTokenProgram: slotAt(fixture, 9).pubkey,
    inputMint: slotAt(fixture, 10).pubkey,
    outputMint: slotAt(fixture, 11).pubkey,
    observationState: slotAt(fixture, 12).pubkey,
  });
}

export function reachableAccountCounts(): Array<number> {
  return [
    build({
      user: key(1),
      ammConfig: key(2),
      pool: key(3),
      inputTokenAccount: key(4),
      outputTokenAccount: key(5),
      inputVault: key(6),
      outputVault: key(7),
      inputTokenProgram: key(8),
      outputTokenProgram: key(9),
      inputMint: key(10),
      outputMint: key(11),
      observationState: key(12),
    }).accountCount,
  ];
}

File: clients/rust/turk-router/tests/common/harness/raydium_cpmm.rs
use solana_pubkey::Pubkey;
use turk_router::venues::raydium_cpmm::{self, RaydiumCpmmAccounts};
use turk_router::VenueWindow;

use crate::common::fixture::Fixture;

pub fn resolve(fixture: &Fixture) -> VenueWindow {
    assert_eq!(fixture.program_id, raydium_cpmm::PROGRAM_ID);
    raydium_cpmm::resolve(RaydiumCpmmAccounts {
        user: fixture.slot(0).pubkey,
        amm_config: fixture.slot(2).pubkey,
        pool: fixture.slot(3).pubkey,
        input_token_account: fixture.slot(4).pubkey,
        output_token_account: fixture.slot(5).pubkey,
        input_vault: fixture.slot(6).pubkey,
        output_vault: fixture.slot(7).pubkey,
        input_token_program: fixture.slot(8).pubkey,
        output_token_program: fixture.slot(9).pubkey,
        input_mint: fixture.slot(10).pubkey,
        output_mint: fixture.slot(11).pubkey,
        observation_state: fixture.slot(12).pubkey,
    })
}

pub fn reachable_account_counts() -> Vec<u8> {
    let key = |byte: u8| Pubkey::new_from_array([byte; 32]);
    vec![raydium_cpmm::resolve(RaydiumCpmmAccounts {
        user: key(1),
        amm_config: key(2),
        pool: key(3),
        input_token_account: key(4),
        output_token_account: key(5),
        input_vault: key(6),
        output_vault: key(7),
        input_token_program: key(8),
        output_token_program: key(9),
        input_mint: key(10),
        output_mint: key(11),
        observation_state: key(12),
    })
    .account_count()]
}
